src/node/stage/approved.rs

```diff
@@ -300,115 +296,273 @@ impl Approved {
+        }
+
+        Ok(())
+    }
+
+    /// Handle message that is "unknown" because we are not in the correct state (e.g. we are adult
+    /// and the message is for elders). We bounce the message to our elders who have more
+    /// information to decide what to do with it.
+    pub fn handle_unknown_message(
+        &self,
+        core: &mut Core,
+        sender: Option<SocketAddr>,
+        msg_bytes: Bytes,
+    ) -> Result<()> {
+        let bounce_msg = Message::single_src(
+            &core.full_id,
+            DstLocation::Direct,
+            None,
+            Variant::BouncedUnknownMessage {
+                message: msg_bytes,
+                parsec_version: self.consensus_engine.parsec_version(),
+            },
+        )?;
+        let bounce_msg = bounce_msg.to_bytes()?;
+
+        // If the message came from one of our elders then bounce it only to them to avoid message
+        // explosion.
+        let our_elder_sender = sender.filter(|sender| {
+            self.shared_state
+                .sections
+                .our_elders()
+                .any(|p2p_node| p2p_node.peer_addr() == sender)
+        });
+        if let Some(sender) = our_elder_sender {
+            core.send_message_to_target(&sender, bounce_msg)
+        } else {
+            self.send_message_to_our_elders(core, bounce_msg)
+        }
+
+        Ok(())
+    }
+
+    pub fn handle_bounced_untrusted_message(
+        &mut self,
+        core: &mut Core,
+        dst_key: Option<bls::PublicKey>,
+        mut bounced_msg: Message,
+    ) -> Result<()> {
+        let dst_key = if let Some(dst_key) = dst_key {
+            dst_key
+        } else {
+            trace!(
+                "Received BouncedUntrustedMessage({:?}) - missing dst key, discarding",
+                bounced_msg,
+            );
+            return Ok(());
+        };
+
+        if let Err(error) = bounced_msg
+            .src
+            .extend_proof(&dst_key, &self.shared_state.our_history)
+        {
+            trace!(
+                "Received BouncedUntrustedMessage({:?}) - extending proof failed, \
+                 discarding: {:?}",
+                bounced_msg,
+                error,
+            );
+            return Ok(());
+        }
+
+        trace!(
+            "Received BouncedUntrustedMessage({:?}) - resending with extended proof",
+            bounced_msg
+        );
+
+        let msg = MessageWithBytes::new(bounced_msg)?;
+        self.relay_message(core, &msg)
+    }
+
+    pub fn handle_bounced_unknown_message(
+        &mut self,
+        core: &mut Core,
+        sender: P2pNode,
+        bounced_msg_bytes: Bytes,
+        sender_parsec_version: u64,
+    ) {
+        if sender_parsec_version >= self.consensus_engine.parsec_version() {
+            trace!(
+                "Received BouncedUnknownMessage({:?}) from {} \
+                 - peer is up to date or ahead of us, discarding",
```

Comment:
there is no early return below this trace log.
not discarding?

Reply:
Good catch!